### DISCUSSION TITLE
[Proposal]: `[DisableImplicitConversion]` on parameters

### DISCUSSION THREAD

**Discussion by Youssef1313** [dotnet/csharplang#7312] [Language Ideas]
# `[DisableImplicitConversion]` on parameters

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Disallows implicit conversions on a specific parameter when doing overload resolution.

## Motivation
[motivation]: #motivation

I came across a case where there is a method that needs to be passed and `object`, where some callers may be passing value types, causing boxing allocations.

The solution to that is to create shared boxes and do it like this:

```csharp
public void M(object o) => ...;
public void M(bool o) => M(o ? Boxes.BoxedTrue : Boxes.BoxedFalse);
public void M(double o) => M(Boxes.Box(o)); // Boxes.Box will return shared box for limited set of doubles. Otherwise it will allocate.
```

The problem with that is now `M(floatValue)` will go to the double overload due to implicit conversion, which is not desired in this case.

This can be avoided by adding more overloads to all known numeric types that are implicitly convertible to double, but this is too much, and doesn't properly handle any user-defined conversion.

This feature can be used by UI Frameworks that are using dependency properties system, such as WPF.

```csharp
public float P
{
    get => (float)GetValue(PProperty);
    // Assume there are two SetValue overloads, one that takes the value as object and other as double.
    // The double parameter will be annotated as DisableImplicitConversion
    // So the following binds to SetValue(DependencyProperty,object)
    set => SetValue(PProperty, value);
}
```

## Detailed design
[design]: #detailed-design

A new attribute is added to the .NET runtime, `DisableImplicitConversionAttribute` which can be added to parameters.

When the compiler does overload resolution, it will not consider implicit conversions for such parameters, and will require exactly the same type.

## Drawbacks
[drawbacks]: #drawbacks

- A library that is using the attribute in public methods may loose the feature guarantees if the consumer is using an older compiler that doesn't support this feature. This could be documented as a limitation of the feature?
- This is achievable by writing an analyzer. Given that usage of this feature might not be common, an analyzer might be a better approach rather than a whole new language feature.

**Comment by Youssef1313** [comment]
Convinced myself that an analyzer would be sufficient and it might not be worth a whole new language feature. Hence I'm leaning towards closing this discussion.

Still would love to hear any other thoughts now covered in the original description.

**Comment by tfenise** [comment]
Why not

```C#
public void M<T>(T o) where T : notnull
{
    if (typeof(T) == typeof(bool)) MCore((bool)(object)o);
    else if (typeof(T) == typeof(double)) MCore((double)(object)o);
    else MCore((object)o);
}

private void MCore(object o) => ...;
private void MCore(bool o) => MCore(o ? Boxes.BoxedTrue : Boxes.BoxedFalse);
private void MCore(double o) => MCore(Boxes.Box(o));
```
